metron-platform/metron-common/src/main/java/org/apache/metron/common/configuration/enrichment/handler/StellarConfig.java

```diff
@@ -142,8 +143,14 @@ else if(kv.getValue() instanceof List) {
   {
 
```

Comment:
Why would this be an either or situation?
What if I have a 'match' expression that uses both _ and other vars?

Reply:
If you are using a `_`, then you cannot make a judgement about which fields you're using, so you have to include all fields.  If you aren't, then you can make some decision and only want to send the fields required to the adapter.  That's the basis for the either/or situation.

Reply:
Well, I added it because I came across it multiple times in data science use-cases.  Specifically, the way that we do model application via MaaS (and other model applications) presumes that the user will filter just the fields into the model, rather than having the model select it.  This puts an undue burden on the user as the model knows which fields it wants and the user may not, so being able to pass the whole message lets the model do the decision.  I think that `_` is a natural way to do that.  To be fair, also, this is a capability inherent to the VariableResolver, not the grammar itself, which I thought was reasonable.  SOme variable resolvers won't support it and some will.

Thoughts?

Reply:
I don't doubt the need or applicability, I guess I'm saying why not just have the _ in the configuration side, and just have the scripts reference the vars by name and not have to MAP_GET()?

That way, the vars are the vars are the vars in Stellar, and the external configuration can have the option for putting all the message in the var namespace.

Reply:
I'm trying to wrap my head around why you need the special handling in the variable resolver.
If one of the vars points to a map, that is it.  You don't need to do any special dance.

You can have the variable context 'builder' ( config whatever ) allow naming the message as a var with a certain name and that is it.

I don't understand why all the extra stuff is required.  IE.  why the variable resolver needs to know it is a map.  It is just a var.

Reply:
I thought about about the variable context builder idea over lunch.  Is what you are proposing to change this so that `MapVariableResolver`, instead of taking a list of maps, we take a Map of maps.  e.g. `MapVariableResolver(Map<String, Map<String, Object>> variables)` and the resolution would check that map first?  So, for instance:
```
MapVariableResolver resolver = new MapVariableResolver(ImmutableMap.of("_", message, "global", globalConfig));
```

Reply:
Something like that, in leu of literal scopes ( which would have a hierarchy ) might work out.  

I'm not trying to shoot down your idea, I just want to understand it, and we approach things differently.

I think in the end we want to get to
interface Scope {
Scope getParent()
Resolve()
Get()
Put()
}

before calling a function/lambda

Scope scope = new Scope(stack.peek())
stack.push(scope)
call function(scope)
stack.pop(scope)

Reply:
So, what you're trying to solve, I think, is scoped variable resolution.  I think this is a bit of a different use-case here and one that doesn't really have an analogue in programming languages at large.  This is, rather, getting a list of the variables and their values within the current scope.  For our purposes, though, we only have one scope, global scope.  I'd consider non-global scope resolution to be out of scope for this.

Ultimately, what I'm arguing for is the map variable resolver to have a special variable that returns the full variable scope. 

Is this becoming clearer or am I spreading more mud than I'm clearing?

Reply:
My impression is that you are creating an implicit scope with the message, maybe that is where we are crossing

Reply:
The PR as it stands does not make any scoping designation beyond global.  Indeed, it is pulling in ALL of the available variables.  My adjustment above adds namespacing (e.g. _ implies the message variables, `global` imples just the global config variables).